ArduPlane: convert parameter TRIM_PITCH_CD to TRIM_PITCH_DEG

timtuxworth opened the pull request:
There is a target to do some refactoring of parameters in the 4.5 release. This is mostly about targeting removing the use of centi-degrees and centimeters in favour of degrees and meters as well as some other parameter cleanups. There is a list documented in this issue: https://github.com/ArduPilot/ardupilot/issues/23580#issuecomment-1520216636

The parameter TRIM_PITCH_CD uses centi degrees. This PR is would replace that with TRIM_PITCH_DEG in degrees. 

I understand now why _DEG is important. Not so much for this parameter, but for other parameters (e.g. Q_ANGLE_MAX) which are in "centi" units but don't have a suffix now, so if we just changed them in situ it would be very dangerously easy to set a new Q_ANGLE_MAX in centi-degrees without realizing it was now degrees. So the only safe way to do this is:

- change Q_ANGLE_MAX to Q_ANGLE_MAX_DEG 
- therefore change all "degrees" parameters being changed from centi units to _DEG 
- therefore as first cab off the rank - TRIM_PITCH_CD becomes TRIM_PITCH_DEG

This PR refactors all the uses of pitch_trim_cd to use pitch_trim scaled down by 100, which almost always simplifies the code and removes floating point operations, since the code needed degrees anyway.

I've added the suggestion from Tridge to create a generic convert_centi_parameter() method on parameters which I copied from some previous code. It does work, but I don't totally understand what it's doing, so please check my work on this.

IamPete1 commented on libraries/AP_Param/AP_Param.cpp:
```diff
@@ -1947,6 +1947,17 @@ void AP_Param::convert_old_parameters(const struct ConversionInfo *conversion_ta
 }
 
+// convert old vehicle parameters to new object parameters with scaling - assumes all parameters will have the same scaling factor
+void AP_Param::convert_old_parameters(const struct ConversionInfo *conversion_table, uint8_t table_size, float scaler, uint8_t flags)
```
```suggestion
void AP_Param::convert_old_parameters_scale(const struct ConversionInfo *conversion_table, uint8_t table_size, float scaler, uint8_t flags)
```

We trying to avoid relying on the argument fingerprint as it can be confusing.

IamPete1 replied:
the existing method should call straight into this one with a 1.0 as the scale.

timtuxworth replied:
Not sure I agree, it's a pretty standard way to write C++, but if that's the standard happy to go with it.

timtuxworth replied:
Done

IamPete1 commented on ArduPlane/Parameters.cpp:
```diff
@@ -1543,6 +1543,17 @@ void Plane::load_parameters(void)
+        AP_Param::convert_old_parameters(&centimeters_conversion_info[0], ARRAY_SIZE(centimeters_conversion_info), 0.02f, 0);
+    }
+    // Create a shim to enable the existing code to work, which should be fixed at a later date
+    g.pitch_trim_cd.set((int16_t)(g.pitch_trim * 100.0f));
```
Can't do this, it would mean you would have to reboot to get the param to work.

timtuxworth replied:
In the dev call on Monday I remember Tridge suggesting this I think:- just changing the parameter name without changing the code. Do you think he missed that this won't work or has some "Tridge" magic idea to make it happen?

IamPete1 replied:
The code change in that case was related to the position controller which uses centimeters internally as well as via the exposed params. This method is fine, it just can't be a one time thing at boot. You could add a function `get_pitch_trim_cm` with does the `pitch_trim * 100.0` then just use it in place of the old param where ever it is used.

timtuxworth replied:
I might as well just fix it properly then.

timtuxworth replied:
I removed the shim and fixed all the usages of pitch_trim_cd - there weren't too many.

IamPete1 commented on ArduPlane/Parameters.cpp:
```diff
@@ -1543,6 +1543,17 @@ void Plane::load_parameters(void)
+        const AP_Param::ConversionInfo centimeters_conversion_info[] {
+            { Parameters::k_param_pitch_trim_cd, 0, AP_PARAM_FLOAT, "PITCH_TRIM" },
+        };
+        AP_Param::convert_old_parameters(&centimeters_conversion_info[0], ARRAY_SIZE(centimeters_conversion_info), 0.02f, 0);
```
```suggestion
        AP_Param::convert_old_parameters(&centimeters_conversion_info[0], ARRAY_SIZE(centimeters_conversion_info), 0.01, 0);
```

IamPete1 reviewed (commented):
I tend to think were better off adding the new params to the end of the g2 group rather than trying to squeeze them in the base table.

rmackay9 commented:
Re TRIM_PITCH vs PITCH_TRIM, I think it depends upon whether you think of the trim as a feature or not.  My guess is that it should remain TRIM_PITCH, TRIM_ROLL and TRIM_YAW (or whatever).

timtuxworth commented:
> Re TRIM_PITCH vs PITCH_TRIM, I think it depends upon whether you think of the trim as a feature or not. My guess is that it should remain TRIM_PITCH, TRIM_ROLL and TRIM_YAW (or whatever).

Yes that's very deliberate. I definitely want to have all the PITCH_ stuff show up together rather than grouping by TRIM as it is now, which I find counterintuitive. Very interested in other peoples thoughts.

timtuxworth commented:
> I tend to think were better off adding the new params to the end of the g2 group rather than trying to squeeze them in the base table.

There is also an aparam.pitch_trim_cd, but I can't figure out where it gets set. (there is no ASCALAR);

IamPete1 commented on libraries/AP_Param/AP_Param.cpp:
```diff
@@ -1947,6 +1948,17 @@ void AP_Param::convert_old_parameters(const struct ConversionInfo *conversion_ta
     flush();
```
You don't need contents of the method anymore if your calling into the other one.

IamPete1 commented on libraries/AP_TECS/AP_TECS.cpp:
```diff
@@ -849,7 +849,7 @@ void AP_TECS::_update_throttle_without_airspeed(int16_t throttle_nudge)
     const float pitch_demand_hpf = _pitch_dem - _pitch_demand_lpf.get();
     _pitch_measured_lpf.apply(_ahrs.pitch, _DT);
-    const float pitch_corrected_lpf = _pitch_measured_lpf.get() - radians(0.01f * (float)aparm.pitch_trim_cd);
+    const float pitch_corrected_lpf = _pitch_measured_lpf.get() - radians(aparm.pitch_trim);
```
Because the old method was broken this is a change in behavior that will need testing.

IamPete1 commented on ArduPlane/Parameters.cpp:
```diff
@@ -1543,6 +1543,17 @@ void Plane::load_parameters(void)
+        };
+        AP_Param::convert_old_parameters_scaled(&centimeters_conversion_info[0], ARRAY_SIZE(centimeters_conversion_info), 0.01f, 0);
+    }
+    // Create a shim to enable the existing code to work, which should be fixed at a later date
+    // g.pitch_trim_cd.set((int16_t)(g.pitch_trim * 100.0f));
```
```suggestion
```

timtuxworth replied:
Done

IamPete1 commented on libraries/AP_Vehicle/AP_FixedWing.h:
```diff
@@ -23,7 +23,7 @@ struct AP_FixedWing {
     AP_Int8  stall_prevention;
     AP_Int16 loiter_radius;
-    AP_Int16 pitch_trim_cd;
+    AP_Int16 pitch_trim;
```
Should be a float.
```suggestion
    AP_Float pitch_trim;
```

timtuxworth replied:
Oops!

IamPete1 commented on ArduPlane/Parameters.cpp:
```diff
@@ -644,14 +644,14 @@ const AP_Param::Info Plane::var_info[] = {
-    // @Range: -4500 4500
+    // @Units: deg
+    // @Range: -45 45
     // @Increment: 10
```
Need to update (or remove) the increment.

timtuxworth replied:
Removed

IamPete1 commented on ArduPlane/Parameters.cpp:
```diff
@@ -1543,6 +1543,17 @@ void Plane::load_parameters(void)
+    // Convert _CM (centimeter) parameters to meters and _CD (centidegrees) parameters to meters
+    {
+        const AP_Param::ConversionInfo centimeters_conversion_info[] {
+            { Parameters::k_param_pitch_trim_cd_old, 0, AP_PARAM_FLOAT, "PITCH_TRIM" },
```
I think the type here is the old param type.
```suggestion
            { Parameters::k_param_pitch_trim_cd_old, 0, AP_PARAM_INT16, "PITCH_TRIM" },
```

timtuxworth replied:
Done

timtuxworth commented:
> I tend to think were better off adding the new params to the end of the g2 group rather than trying to squeeze them in the base table.

I can't figure out how to make this a g2 parameter now it's in aparm. Is there an equivalent g2 macro to ASCALAR?

OTOH- the base table is already > 256 elements, so does it really matter?

IamPete1 commented on ArduPlane/Attitude.cpp:
```diff
@@ -209,7 +209,7 @@ float Plane::stabilize_pitch_get_pitch_out()
 #endif
 
-    int32_t demanded_pitch = nav_pitch_cd + g.pitch_trim_cd + SRV_Channels::get_output_scaled(SRV_Channel::k_throttle) * g.kff_throttle_to_pitch;
+    int32_t demanded_pitch = nav_pitch_cd + uint32_t(aparm.pitch_trim * 100) + SRV_Channels::get_output_scaled(SRV_Channel::k_throttle) * g.kff_throttle_to_pitch;
```
```suggestion
    int32_t demanded_pitch = nav_pitch_cd + int32_t(aparm.pitch_trim * 100) + SRV_Channels::get_output_scaled(SRV_Channel::k_throttle) * g.kff_throttle_to_pitch;
```

timtuxworth replied:
Thank you!

tridge commented on ArduPlane/Parameters.cpp:
```diff
@@ -644,14 +644,13 @@ const AP_Param::Info Plane::var_info[] = {
+    // @Units: deg
+    // @Range: -45 45
+    // @User: Standard
+    ASCALAR(pitch_trim,        "PITCH_TRIM",  0.0f),
```
I would greatly prefer TRIM_PITCH_DEG so users looking for the old parameter can find the new one easily

timtuxworth replied:
I thought part of the point of getting rid of the "CD" was that the units are already in the metadata. Users can easily find the parameter by typing "TRIM" or "PITCH" in the search box.

timtuxworth replied:
Oh and I just noticed this comment in the related issue - maybe it would be better to use PTCH because it's the most consistently used tag for "PITCH" which often needs to be abbreviated. https://github.com/ArduPilot/ardupilot/issues/23580#issuecomment-1519206557

IamPete1 replied:
I agree with Tim, I don't think we need to add deg for si units. Adding deg also sets up a pattern we may not be able to follow as deg is one character longer than cd.

We could add a magnitude arming check, anything above say 10 we could assume the user copied the cd value.

tridge replied:
that shouldn't stop us adding _DEG when we can. It will save users a lot of pain and most params can get the _DEG

timtuxworth replied:
There are already at least 13 PITCH parameters that are in degrees, but do NOT have "DEG" in the name, and none that do. Not to mention all the YAW and ROLL parameters that are in degrees with no "DEG" at the end.

IMO it's inconsistency that is most painful to users, and again - it's in the metadata, which is displayed in both Mission Planner and QGroundControl.

timtuxworth replied:
Just noticed this one: Q_TRIM_PITCH: Quadplane AHRS trim pitch (in degrees, no _DEG) so TRIM_PITCH would be consistent with this. without it we will have

- TRIM_PITCH_DEG (in degrees)
- Q_TRIM_PITCH (in degrees)

timtuxworth replied:
Ok - I get it _DEG is necessary. So I've reworked this now.

tridge commented on ArduPlane/Parameters.cpp:
```diff
@@ -1543,6 +1542,15 @@ void Plane::load_parameters(void)
+    // PARAMETER_CONVERSION - Added: Dec 2023
+    // Convert _CM (centimeter) parameters to meters and _CD (centidegrees) parameters to meters
+    {
+        const AP_Param::ConversionInfo centimeters_conversion_info[] {
```
we need to create a special method for this in AP_Param, so we can do:
  myparam.convert_centi_parameter();
and it is done

timtuxworth replied:
Not really. That way you have to write a loop. This way you can pass in a whole table of parameters using the existing function - which is what I'm expecting next ...

```
        const AP_Param::ConversionInfo centimeters_conversion_info[] {
            { Parameters::k_param_pitch_trim_cd_old, 0, AP_PARAM_INT16, "PITCH_TRIM" },
            { Parameters::k_param_lim_pitch_max_cd_old, 0, AP_PARAM_INT16, "PITCH_LIM_MAX" },
            { Parameters::k_param_lim_pitch_min_cd_old, 0, AP_PARAM_INT16, "PITCH_LIM_MIN" },
            { Parameters::k_param_lim_roll_old, 0, AP_PARAM_INT16, "ROLL_LIM" },
            { Parameters::k_param_airspeed_cruise_old, 0, AP_PARAM_INT16, "AIRSPEED_CRUISE" },
        };
        AP_Param::convert_old_parameters_scaled(&centimeters_conversion_info[0], ARRAY_SIZE(centimeters_conversion_info), 0.01f, 0);

```

timtuxworth replied:
> we need to create a special method for this in AP_Param, so we can do: myparam.convert_centi_parameter(); and it is done

I took a stab at it by copying the existing convert_parameter_width() method in AP_Param, it seems to work, but it's very low level so I wonder if you could take a look. I had to pass in the old type to get it to work.

timtuxworth replied:
Tested in SITL

tridge commented on ArduPlane/Parameters.h:
```diff
@@ -56,7 +56,7 @@ class Parameters {
         k_param_auto_trim      = 10, // unused
         k_param_log_bitmask_old,  // unused
-        k_param_pitch_trim_cd,
+        k_param_pitch_trim_cd_old,  // replaced by pitch_trim
```
we should keep the same ID

timtuxworth replied:
I'm good with that - I didn't know it would work.

timtuxworth replied:
fixed

tridge commented on ArduPlane/Parameters.h:
```diff
@@ -437,7 +441,6 @@ class Parameters {
     AP_Int16 dspoiler_rud_rate;
     AP_Int32 log_bitmask;
     AP_Int32 RTL_altitude_cm;
-    AP_Int16 pitch_trim_cd;
```
it should just change to AP_Float

timtuxworth replied:
It had already been partly moved to AP_FixedWing;:aparm .. I'm just completing the process. See the above discussion about #22191

tridge replied:
it doesn't need to be moved

timtuxworth replied:
fixed

timtuxworth commented:
PITCH_TRIM auto converted from TRIM_PITCH_CD = 250 on a Durandal
<img width="314" alt="Screenshot 2023-12-16 at 6 54 31 PM" src="https://github.com/ArduPilot/ardupilot/assets/5266784/99ac75b4-b685-4182-9880-6a7748deb137">
<img width="536" alt="Screenshot 2023-12-16 at 6 54 48 PM" src="https://github.com/ArduPilot/ardupilot/assets/5266784/00f32541-d281-4178-8360-2838d01be685">

tridge commented:
> I tend to think were better off adding the new params to the end of the g2 group rather than trying to squeeze them in the base table.

we don't need to move them at all, they can keep the same ID

davidbuzz commented:
I think this is as close to what tridge wanted, based on today's discussion:  ( $ means end-of-string) 
I took a full list of all possible plane and copter params, and wroe a quick tool to replace  _CD or CD$ with _DEG or DEG, and then _CM or CM$ with _M or M, and here's the results:
PLANE , RNGFND1_MIN_CM, LEN, 14, newname, RNGFND1_MIN_M, newlen, 13, 
PLANE , RNGFND1_MAX_CM, LEN, 14, newname, RNGFND1_MAX_M, newlen, 13, 
PLANE , RNGFND2_MIN_CM, LEN, 14, newname, RNGFND2_MIN_M, newlen, 13, 
PLANE , RNGFND2_MAX_CM, LEN, 14, newname, RNGFND2_MAX_M, newlen, 13, 
PLANE , RNGFND3_MIN_CM, LEN, 14, newname, RNGFND3_MIN_M, newlen, 13, 
PLANE , RNGFND3_MAX_CM, LEN, 14, newname, RNGFND3_MAX_M, newlen, 13, 
PLANE , RNGFND4_MIN_CM, LEN, 14, newname, RNGFND4_MIN_M, newlen, 13, 
PLANE , RNGFND4_MAX_CM, LEN, 14, newname, RNGFND4_MAX_M, newlen, 13, 
PLANE , RNGFND5_MIN_CM, LEN, 14, newname, RNGFND5_MIN_M, newlen, 13, 
PLANE , RNGFND5_MAX_CM, LEN, 14, newname, RNGFND5_MAX_M, newlen, 13, 
PLANE , RNGFND6_MIN_CM, LEN, 14, newname, RNGFND6_MIN_M, newlen, 13, 
PLANE , RNGFND6_MAX_CM, LEN, 14, newname, RNGFND6_MAX_M, newlen, 13, 
PLANE , RNGFND7_MIN_CM, LEN, 14, newname, RNGFND7_MIN_M, newlen, 13, 
PLANE , RNGFND7_MAX_CM, LEN, 14, newname, RNGFND7_MAX_M, newlen, 13, 
PLANE , RNGFND8_MIN_CM, LEN, 14, newname, RNGFND8_MIN_M, newlen, 13, 
PLANE , RNGFND8_MAX_CM, LEN, 14, newname, RNGFND8_MAX_M, newlen, 13, 
PLANE , RNGFND9_MIN_CM, LEN, 14, newname, RNGFND9_MIN_M, newlen, 13, 
PLANE , RNGFND9_MAX_CM, LEN, 14, newname, RNGFND9_MAX_M, newlen, 13, 
PLANE , RNGFNDA_MIN_CM, LEN, 14, newname, RNGFNDA_MIN_M, newlen, 13, 
PLANE , RNGFNDA_MAX_CM, LEN, 14, newname, RNGFNDA_MAX_M, newlen, 13, 
PLANE , LIM_ROLL_CD, LEN, 11, newname, LIM_ROLL_DEG, newlen, 12, 
PLANE , TRIM_ARSPD_CM, LEN, 13, newname, TRIM_ARSPD_M, newlen, 12, 
PLANE , MIN_GNDSPD_CM, LEN, 13, newname, MIN_GNDSPD_M, newlen, 12, 
PLANE , TRIM_PITCH_CD, LEN, 13, newname, TRIM_PITCH_DEG, newlen, 14, 
PLANE , ALT_HOLD_FBWCM, LEN, 14, newname, ALT_HOLD_FBWM, newlen, 13, 
PLANE , LAND_PITCH_CD, LEN, 13, newname, LAND_PITCH_DEG, newlen, 14, 
PLANE , RNGFND1_MIN_CM, LEN, 14, newname, RNGFND1_MIN_M, newlen, 13, 
PLANE , RNGFND1_MAX_CM, LEN, 14, newname, RNGFND1_MAX_M, newlen, 13, 
PLANE , RNGFND2_MIN_CM, LEN, 14, newname, RNGFND2_MIN_M, newlen, 13, 
PLANE , RNGFND2_MAX_CM, LEN, 14, newname, RNGFND2_MAX_M, newlen, 13, 
PLANE , RNGFND3_MIN_CM, LEN, 14, newname, RNGFND3_MIN_M, newlen, 13, 
PLANE , RNGFND3_MAX_CM, LEN, 14, newname, RNGFND3_MAX_M, newlen, 13, 
PLANE , RNGFND4_MIN_CM, LEN, 14, newname, RNGFND4_MIN_M, newlen, 13, 
PLANE , RNGFND4_MAX_CM, LEN, 14, newname, RNGFND4_MAX_M, newlen, 13, 
PLANE , RNGFND5_MIN_CM, LEN, 14, newname, RNGFND5_MIN_M, newlen, 13, 
PLANE , RNGFND5_MAX_CM, LEN, 14, newname, RNGFND5_MAX_M, newlen, 13, 
PLANE , RNGFND6_MIN_CM, LEN, 14, newname, RNGFND6_MIN_M, newlen, 13, 
PLANE , RNGFND6_MAX_CM, LEN, 14, newname, RNGFND6_MAX_M, newlen, 13, 
PLANE , RNGFND7_MIN_CM, LEN, 14, newname, RNGFND7_MIN_M, newlen, 13, 
PLANE , RNGFND7_MAX_CM, LEN, 14, newname, RNGFND7_MAX_M, newlen, 13, 
PLANE , RNGFND8_MIN_CM, LEN, 14, newname, RNGFND8_MIN_M, newlen, 13, 
PLANE , RNGFND8_MAX_CM, LEN, 14, newname, RNGFND8_MAX_M, newlen, 13, 
PLANE , RNGFND9_MIN_CM, LEN, 14, newname, RNGFND9_MIN_M, newlen, 13, 
PLANE , RNGFND9_MAX_CM, LEN, 14, newname, RNGFND9_MAX_M, newlen, 13, 
PLANE , RNGFNDA_MIN_CM, LEN, 14, newname, RNGFNDA_MIN_M, newlen, 13, 
PLANE , RNGFNDA_MAX_CM, LEN, 14, newname, RNGFNDA_MAX_M, newlen, 13, 
PLANE , SOAR_POLAR_CD0, LEN, 14, newname, SOAR_POLAR_DEG0, newlen, 15, 
_CD count: 4
_CM count: 43

davidbuzz commented:
as discussed, the above list also doesn't include params that have centimeters in them, but don't end in _CM

davidbuzz commented:
This was force-pushed, so just noting that it still isn't aligned with the renaming that tridge said he prefers in the dev call. TRIM_PITCH_CD => TRIM_PITCH_DEG , so I suspect it won't be mergable.

timtuxworth commented:
> This was force-pushed, so just noting that it still isn't aligned with the renaming that tridge said he prefers in the dev call. TRIM_PITCH_CD => TRIM_PITCH_DEG , so I suspect it won't be mergable.

I was a bit premature with my push, I need to do more testing.

tridge commented:
@timtuxworth I've rebased and cleaned this up for TRIM_PITCH_DEG

timtuxworth commented:
> @timtuxworth I've rebased and cleaned this up for TRIM_PITCH_DEG

Thanks @tridge  I was just working on this today, I'll try to finish it up in the next day or two.

tridge commented:
replaced by #26025
